### PR TITLE
Download and Install Docker

### PR DESCRIPTION
**Pull request by ravipal**
When docker is not installed, earlier the docker view shows 'Install Docker' that will open up a web page. Now clicking the node will start the download and initiate the installation.

Fixes: #1785

### CODE REVIEW

File: src/commands/installDocker.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import { IActionContext } from 'vscode-azureextensionui';
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import LocalOSProvider from '../utils/LocalOSProvider';
+import { DockerInstallerBase } from './dockerInstallerBase';
+
+class WindowsDockerInstaller extends DockerInstallerBase {
+    protected downloadUrl: string = 'https://aka.ms/download-docker-windows-vscode';
+    protected fileExtension: string = 'exe';
+    protected installationMessage: string = localize('vscode-docker.commands.WindowsDockerInstaller.installationMessage', 'Installer is launched. Please follow the prompt and complet the installation and start the Docker Desktop.');
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:15]:
```suggestion
    protected installationMessage: string = localize('vscode-docker.commands.WindowsDockerInstaller.installationMessage', 'Installer is launched. Please follow the prompts to complete the installation, and then start the Docker Desktop.');
```

File: src/commands/installDocker.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    protected getInstallCommand(fileName: string): string {
+        // Windows require double quote.
+        return `"${fileName}"`;
+    }
+}
+
+class MacDockerInstaller extends DockerInstallerBase {
+    protected downloadUrl: string = 'https://aka.ms/download-docker-mac-vscode';
+    protected fileExtension: string = 'dmg';
+    protected installationMessage: string = localize('vscode-docker.commands.MacDockerInstaller.installationMessage', 'Installer is launched. Please follow the prompt and complete the installation and start the Docker Desktop.');
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:25]:
```suggestion
    protected installationMessage: string = localize('vscode-docker.commands.MacDockerInstaller.installationMessage', 'Installer is launched. Please follow the prompts to complete the installation, and then start the Docker Desktop.');
```

File: src/commands/installDocker.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    protected getInstallCommand(fileName: string): string {
+        return `Chmod +x '${fileName}' && open '${fileName}'`;
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:27]:
```suggestion
        return `chmod +x '${fileName}' && open '${fileName}'`;
```

File: src/commands/installDocker.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    }
+}
+
+class LinuxDockerInstaller extends DockerInstallerBase {
+    protected downloadUrl: string = 'https://aka.ms/download-docker-linux-vscode';
+    protected fileExtension: string = 'sh';
+    protected installationMessage: string = localize('vscode-docker.commands.LinuxDockerInstaller.installationMessage', 'Please follow the prompt in terminal window and complete the installation and start the Docker.');
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:34]:
```suggestion
    protected installationMessage: string = localize('vscode-docker.commands.LinuxDockerInstaller.installationMessage', 'Please follow the prompt in terminal window to complete the installation, and then start Docker.');
```

File: src/commands/installDocker.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    protected getInstallCommand(fileName: string): string {
+        return `chmod +x '${fileName}' && sh '${fileName}'`;
+    }
+
+    protected async install(fileName: string): Promise<void> {
+        const terminal = ext.terminalProvider.createTerminal(localize('vscode-docker.commands.LinuxDockerInstaller.terminalTitle', 'Docker Install.'));
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:40]:
```suggestion
        const terminal = ext.terminalProvider.createTerminal(localize('vscode-docker.commands.DockerInstaller.terminalTitle', 'Docker Install'));
```

File: src/commands/installDocker.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        const command = this.getInstallCommand(fileName);
+        terminal.sendText(command);
+        terminal.show();
+        return;
+    }
+}
+
+export async function installDocker(context: IActionContext): Promise<void> {
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:48]: I don't have any specific ideas, but is there any extra telemetry properties we want to attach to `context.telemetry`?

File: src/commands/installDocker.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IActionContext } from 'vscode-azureextensionui';
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import LocalOSProvider from '../utils/LocalOSProvider';
+import { DockerInstallerBase } from './dockerInstallerBase';
+
+class WindowsDockerInstaller extends DockerInstallerBase {
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:12]: IMO these three classes make more sense to be in the same file as `DockerInstallerBase`, or otherwise files of their own, rather than in here.

File: src/commands/dockerInstallerBase.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:1]: Change file name to `DockerInstallerBase.ts` (classes get capitalized file names, at least in this repo).

File: src/commands/dockerInstallerBase.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import LocalOSProvider from '../utils/LocalOSProvider';
+
+export abstract class DockerInstallerBase {
+    protected abstract downloadUrl: string;
+    protected abstract fileExtension: string;
+    protected abstract installationMessage: string;
+    protected abstract getInstallCommand(fileName: string): string;
+
+    public async downloadAndInstallDocker(): Promise<void> {
+        const confirmInstall: string = localize('vscode-docker.commands.DockerInstallerBase.confirm', 'Are you sure you want to install docker on this machine.');
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:22]:
```suggestion
        const confirmInstall: string = localize('vscode-docker.commands.DockerInstallerBase.confirm', 'Are you sure you want to install Docker on this machine?');
```

File: src/commands/dockerInstallerBase.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        const installTitle: string = localize('vscode-docker.commands.DockerInstallerBase.install', 'Install');
+        const downloadingMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloading', 'Downloading Docker installer...');
+        const downloadCompleteMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloadCompleteMessage', 'Download completed');
+        let downlaodedFileName: string;
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:26]:
```suggestion
        let downloadedFileName: string;
```

File: src/commands/dockerInstallerBase.ts [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+        // no need to check result - cancel will throw a UserCancelledError
+        await ext.ui.showWarningMessage(confirmInstall, { modal: true }, { title: installTitle });
+
+        await vscode.window.withProgress(
+            { location: vscode.ProgressLocation.Notification, title: downloadingMessage },
+            async () => {
+                downlaodedFileName = await this.downloadInstaller();
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:34]:
```suggestion
                downloadedFileName = await this.downloadInstaller();
```

File: src/commands/dockerInstallerBase.ts [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+    public async downloadAndInstallDocker(): Promise<void> {
+        const confirmInstall: string = localize('vscode-docker.commands.DockerInstallerBase.confirm', 'Are you sure you want to install docker on this machine.');
+        const installTitle: string = localize('vscode-docker.commands.DockerInstallerBase.install', 'Install');
+        const downloadingMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloading', 'Downloading Docker installer...');
+        const downloadCompleteMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloadCompleteMessage', 'Download completed');
+        let downlaodedFileName: string;
+
+        // no need to check result - cancel will throw a UserCancelledError
+        await ext.ui.showWarningMessage(confirmInstall, { modal: true }, { title: installTitle });
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:29]: Make sure @BigMorty is OK with a modal dialog, they are very annoying.

File: src/commands/dockerInstallerBase.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+        await vscode.window.withProgress(
+            { location: vscode.ProgressLocation.Notification, title: downloadingMessage },
+            async () => {
+                downlaodedFileName = await this.downloadInstaller();
+                // eslint-disable-next-line @typescript-eslint/no-floating-promises
+                vscode.window.showInformationMessage(downloadCompleteMessage);
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:36]: Do we need another toast to say download is complete? IMO when the "Downloading Docker installer..." progress toast disappears it's pretty clear that download is done. Also, immediately after this we show the installation message, so I'd rather just remove this toast entirely.

File: src/commands/dockerInstallerBase.ts [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+            }
+        );
+
+        const command = this.getInstallCommand(downlaodedFileName);
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:40]:
```suggestion
        const command = this.getInstallCommand(downloadedFileName);
```

File: src/commands/dockerInstallerBase.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        vscode.window.showInformationMessage(this.installationMessage);
+        await this.install(downlaodedFileName, command);
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:43]:
```suggestion
        await this.install(downloadedFileName, command);
```

File: src/commands/dockerInstallerBase.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+        await streamToFile(this.downloadUrl, fileName);
+        return fileName;
+    }
+
+    protected async install(fileName: string, cmd: string): Promise<void> {
+        const fsProvider = new LocalFileSystemProvider();
+        try {
+            const processProvider = new ChildProcessProvider();
+            ext.outputChannel.appendLine(localize('vscode-docker.commands.DockerInstallerBase.downloadCompleteMessage', 'Executing command {0}', cmd));
+            await processProvider.exec(cmd, {});
```
Comment (bwateratmsft) [inline src/commands/dockerInstallerBase.ts:60]: Use the terminal for all OSs, don't spawn the process directly.

Comment (bwateratmsft) [reply src/commands/dockerInstallerBase.ts:60]: Otherwise, at least use `spawnAsync` directly instead of `ChildProcessProvider`.

File: src/utils/httpRequest.ts [hunk 17]
```diff
@@ -55,3 +57,29 @@ export async function httpsRequestBinary(opts: https.RequestOptions | string): P
         req.on('error', reject);
     });
 }
+
+export async function streamToFile(downloadUrl: string, fileName: string): Promise<void> {
```
[the remark this answers is not in the document]
Comment (ravipal) [reply src/utils/httpRequest.ts:61]: Nope that didn't work. Also that method is not efficient for large files. Streaming is the way to go.

File: src/commands/dockerInstallerBase.ts (hunk 12 above)
Comment (ravipal) [reply src/commands/dockerInstallerBase.ts:29]: Here the modal dialog make sense as we need user input to proceed. @BigMorty reopen this comment, if you don't agree.

File: src/commands/dockerInstallerBase.ts (hunk 13 above)
Comment (ravipal) [reply src/commands/dockerInstallerBase.ts:36]: @BigMorty, Your call.

File: src/commands/dockerInstallerBase.ts (hunk 16 above)
Comment (ravipal) [reply src/commands/dockerInstallerBase.ts:60]: We decided to user Terminal only where we need user input.
Also what is the reasoning behind not using ChildProcessProvider.

Comment (bwateratmsft) [reply src/commands/dockerInstallerBase.ts:60]: Eventually the coreclr things are going to go away or get refactored. I don't want to keep adding to the amount of work that's going to take. `ChildProcessProvider` just calls `spawnAsync` underneath it.

File: src/commands/installDocker.ts (hunk 5 above)
Comment (ravipal) [reply src/commands/installDocker.ts:40]: This is defined in LinuxDockerInstaller so using that in the ID.

File: src/commands/installDocker.ts (hunk 6 above)
Comment (ravipal) [reply src/commands/installDocker.ts:48]: We need OS and success and failure state, which are captured already.

File: src/commands/dockerInstallerBase.ts (hunk 13 above)
Comment (bwateratmsft) [reply src/commands/dockerInstallerBase.ts:36]: Users have repeatedly stated that toasts are annoying. We should minimize them.

File: src/commands/installDocker.ts (hunk 7 above)
Comment (ravipal) [reply src/commands/installDocker.ts:12]: I prefer to keep the base class separate. These 3 classes don't have much logic other than providing the data. I think having it here will give you full picture. Having a separate file each one will be an overkill.

Comment (bwateratmsft) [reply src/commands/installDocker.ts:12]: In that case they should move to DockerInstallerBase.

File: src/commands/installDocker.ts [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    protected downloadUrl: string = 'https://aka.ms/download-docker-windows-vscode';
+    protected fileExtension: string = 'exe';
+    protected installationMessage: string = localize('vscode-docker.commands.WindowsDockerInstaller.installationMessage', 'Installer is launched. Please follow the prompts to complete the installation, and then start the Docker Desktop.');
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:15]:
```suggestion
    protected installationMessage: string = localize('vscode-docker.commands.WindowsDockerInstaller.installationMessage', 'Installer is launched. Please follow the prompts to complete the installation, and then start Docker Desktop.');
```
Sorry, I missed this in the first pass.

File: src/commands/installDocker.ts [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    protected getInstallCommand(fileName: string): string {
+        // Windows require double quote.
+        return `"${fileName}"`;
+    }
+}
+
+class MacDockerInstaller extends DockerInstallerBase {
+    protected downloadUrl: string = 'https://aka.ms/download-docker-mac-vscode';
+    protected fileExtension: string = 'dmg';
+    protected installationMessage: string = localize('vscode-docker.commands.MacDockerInstaller.installationMessage', 'Installer is launched. Please follow the prompts to complete the installation, and then start the Docker Desktop.');
```
Comment (bwateratmsft) [inline src/commands/installDocker.ts:25]:
```suggestion
    protected installationMessage: string = localize('vscode-docker.commands.MacDockerInstaller.installationMessage', 'Installer is launched. Please follow the prompts to complete the installation, and then start Docker Desktop.');
```
(Also missed this)

File: src/commands/dockerInstallerBase.ts [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        vscode.window.showInformationMessage(this.installationMessage);
+        await this.install(downloadedFileName, command);
+    }
+
+    private async downloadInstaller(): Promise<string> {
+        const osProvider = new LocalOSProvider();
+        const processProvider = new ChildProcessProvider();
+        const tempFileProvider = new OSTempFileProvider(osProvider, processProvider);
+        let fileName = tempFileProvider.getTempFilename('docker', this.fileExtension);
```
Comment (haniamr) [inline src/commands/dockerInstallerBase.ts:50]: `const`?

File: src/commands/installDocker.ts (hunk 19 above)
Comment (haniamr) [inline src/commands/installDocker.ts:25]: There aren't really any prompts in Mac, it's just the drag-and-drop part. Do we want to reword this to make it more clear to the user?

File: src/commands/installDocker.ts [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+    protected getInstallCommand(fileName: string): string {
+        return `chmod +x '${fileName}' && open '${fileName}'`;
```
Comment (haniamr) [inline src/commands/installDocker.ts:27]: Is that required in Mac (the chmod part)? I tried with just the "open" command and it works fine.

File: src/commands/installDocker.ts [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+    }
+}
+
+class LinuxDockerInstaller extends DockerInstallerBase {
+    protected downloadUrl: string = 'https://aka.ms/download-docker-linux-vscode';
+    protected fileExtension: string = 'sh';
+    protected installationMessage: string = localize('vscode-docker.commands.LinuxDockerInstaller.installationMessage', 'Please follow the prompt in terminal window to complete the installation, and then start Docker.');
```
Comment (haniamr) [inline src/commands/installDocker.ts:34]: `Docker Desktop`?

File: src/commands/installDocker.ts (hunk 7 above)
Comment (haniamr) [reply src/commands/installDocker.ts:12]: I agree, they should either be in separate files or the base class file.

File: src/commands/dockerInstallerBase.ts (hunk 16 above)
Comment (ravipal) [reply src/commands/dockerInstallerBase.ts:60]: I didn't know that we have a plan to get rid of ChildProcessProvider. I will avoid using it then.

File: src/commands/installDocker.ts (hunk 22 above)
Comment (bwateratmsft) [reply src/commands/installDocker.ts]: Not on Linux. It's "Docker Community Edition" or something like that.

File: src/commands/installDocker.ts (hunk 19 above)
Comment (bwateratmsft) [reply src/commands/installDocker.ts]: It is something that could change depending on what Docker does with their installer / dmg in the future, so it's hard to say.

File: src/commands/dockerInstallerBase.ts (hunk 16 above)
Comment (bwateratmsft) [reply src/commands/dockerInstallerBase.ts]: Yeah. Definitely a lot of it will be reused/refactored, though.

File: src/commands/dockerInstallerBase.ts (hunk 13 above)
Comment (ravipal) [reply src/commands/dockerInstallerBase.ts]: Agree, that the 'Downloaded' is unnecessary. Removed it now. Also sent email to get Kaitlin (UX) input, based on her input, I will adjust.

File: src/commands/dockerInstaller.ts [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+    public async downloadAndInstallDocker(): Promise<void> {
+        const confirmInstall: string = localize('vscode-docker.commands.DockerInstallerBase.confirm', 'Are you sure you want to install Docker on this machine?');
+        const installTitle: string = localize('vscode-docker.commands.DockerInstallerBase.install', 'Install');
+        const downloadingMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloading', 'Downloading Docker installer...');
+        let downloadedFileName: string;
+
+        // no need to check result - cancel will throw a UserCancelledError
+        await ext.ui.showWarningMessage(confirmInstall, { modal: true }, { title: installTitle });
+
+        await vscode.window.withProgress(
```
Comment (bwateratmsft) [inline src/commands/dockerInstaller.ts:31]: It's not imperative but another way you can do this is like this:

```typescript
const downloadedFileName = await vscode.window.withProgress( { ... }, () => this.downloadInstaller() );
```

(There may need to be +/- some `async`/`await` in there, but in general `withProgress` can return the inner value.)

File: src/commands/installDocker.ts (hunk 21 above)
Comment (ravipal) [reply src/commands/installDocker.ts]: I added it based on what Devin ran into. He got access denied or something like that. I think it is safe to add execute permission.

File: src/commands/installDocker.ts (hunk 19 above)
Comment (ravipal) [reply src/commands/installDocker.ts]: Not sure what the right thing to call. I used prompt to indicate the GUI shown to user. Any suggestions?

Comment (bwateratmsft) [reply src/commands/installDocker.ts]: I guess I'm fine with the current verbage. @BigMorty?

File: src/commands/installDocker.ts (hunk 22 above)
Comment (bwateratmsft) [reply src/commands/installDocker.ts]: (Additionally, there's about a billion different ways to set up Docker on Linux and have something that works)
